File: apps/mobile/client-apple/VettaKit/Sources/VettaKit/Protocol/RemoteAPI.swift
import Foundation

/// Payload contracts carried inside requests, responses and events (port of `api.ts`).
/// Readers are tolerant: unknown fields are dropped and malformed entries skipped.
public enum RemoteSessionStatus: String, Codable, Sendable, CaseIterable {
	case idle
	case running
	case thinking
	case waitingInput = "waiting_input"
	case completed
	case error
	case aborted

	public var isActive: Bool { self == .running || self == .thinking || self == .waitingInput }
}

public struct RemoteProjectSummary: Equatable, Codable, Sendable {
	public var cwd: String
	public var name: String
	/// "conversation" for the desktop's project-less chats, otherwise "project".
	public var kind: String
	public var sessionCount: Double

	public init(cwd: String, name: String, kind: String, sessionCount: Double) {
		self.cwd = cwd
		self.name = name
		self.kind = kind
		self.sessionCount = sessionCount
	}

	public var isConversation: Bool { kind == "conversation" }
}

public struct RemoteSessionSummary: Equatable, Codable, Sendable, Identifiable {
	public var id: String
	public var projectCwd: String
	public var projectName: String
	public var title: String
	public var preview: String?
	public var updatedAt: Double
	public var status: RemoteSessionStatus
	/// True when the desktop currently holds a live runtime instance for it.
	public var live: Bool

	public init(id: String, projectCwd: String, projectName: String, title: String, preview: String? = nil, updatedAt: Double, status: RemoteSessionStatus, live: Bool) {
		self.id = id
		self.projectCwd = projectCwd
		self.projectName = projectName
		self.title = title
		self.preview = preview
		self.updatedAt = updatedAt
		self.status = status
		self.live = live
	}
}

public struct RemoteQuestionOption: Equatable, Codable, Sendable {
	public var label: String
	public var description: String
}

public struct RemoteQuestionItem: Equatable, Codable, Sendable {
	public var question: String
	public var header: String
	public var options: [RemoteQuestionOption]
	public var multiSelect: Bool
}

public struct RemoteQuestionRequest: Equatable, Codable, Sendable {
	public var requestId: String
	public var questions: [RemoteQuestionItem]
}

public struct RemoteQuestionAnswer: Equatable, Sendable {
	public var question: String
	public var answers: [String]

	public init(question: String, answers: [String]) {
		self.question = question
		self.answers = answers
	}

	var json: JSONValue { ["question": .string(question), "answers": .array(answers.map(JSONValue.string))] }
}

public struct RemoteSessionError: Equatable, Codable, Sendable {
	public var code: String
	public var message: String
}

public struct RemoteSessionState: Equatable, Codable, Sendable {
	public var status: RemoteSessionStatus
	public var detail: String?
	public var model: String?
	public var contextPercent: Double?
	public var error: RemoteSessionError?
	/// Present while the desktop waits for an answer that the phone may give.
	public var pendingQuestion: RemoteQuestionRequest?

	public init(status: RemoteSessionStatus, detail: String? = nil, model: String? = nil, contextPercent: Double? = nil, error: RemoteSessionError? = nil, pendingQuestion: RemoteQuestionRequest? = nil) {
		self.status = status
		self.detail = detail
		self.model = model
		self.contextPercent = contextPercent
		self.error = error
		self.pendingQuestion = pendingQuestion
	}
}

public enum RemoteToolPhase: String, Sendable {
	case generating, started, updated, phase, completed, failed
}

public struct RemoteToolEvent: Equatable, Sendable {
	public var toolCallId: String
	public var toolName: String
	public var phase: RemoteToolPhase
	public var args: String?
	public var result: String?
	public var label: String?
	public var durationMs: Double?

	public init(toolCallId: String, toolName: String, phase: RemoteToolPhase, args: String? = nil, result: String? = nil, label: String? = nil, durationMs: Double? = nil) {
		self.toolCallId = toolCallId
		self.toolName = toolName
		self.phase = phase
		self.args = args
		self.result = result
		self.label = label
		self.durationMs = durationMs
	}
}

public enum RemoteMessageEvent: Equatable, Sendable {
	case user(text: String, at: Double)
	case assistantDelta(String)
	case thinkingDelta(String)
	case turnEnd(at: Double)
}

public struct RemoteToolCallSummary: Equatable, Sendable {
	public var toolCallId: String
	public var toolName: String
	public var args: String?
	public var result: String?
	public var isError: Bool
	public var durationMs: Double?

	public init(toolCallId: String, toolName: String, args: String? = nil, result: String? = nil, isError: Bool = false, durationMs: Double? = nil) {
		self.toolCallId = toolCallId
		self.toolName = toolName
		self.args = args
		self.result = result
		self.isError = isError
		self.durationMs = durationMs
	}
}

public enum RemoteTranscriptEntry: Equatable, Sendable {
	case user(id: String, text: String, at: Double?)
	case assistant(id: String, text: String, thinking: String?, toolCalls: [RemoteToolCallSummary], at: Double?, error: String?)
	case marker(id: String, text: String, at: Double?)
}

public struct RemoteDeviceStatus: Equatable, Sendable {
	public var deviceName: String
	public var osLabel: String?
	public var lanEndpoints: [String]
	public var relayEnabled: Bool
	public var runningSessionCount: Double
}

/// Sealed follow-up to a manual pairing approval; carries the long-lived credential.
public struct RemoteDevicePaired: Equatable, Sendable {
	public var pairingId: String
	public var mobileSecret: String
	public var desktopName: String
	public var lanEndpoints: [String]
	public var relayBaseUrl: String?
}

public enum RemoteAPI {
	public static func readSessionStatus(_ value: JSONValue?) -> RemoteSessionStatus {
		value?.stringValue.flatMap(RemoteSessionStatus.init(rawValue:)) ?? .idle
	}

	public static func readSessionSummary(_ value: JSONValue?) -> RemoteSessionSummary? {
		guard let value, value.isObject,
		      let id = nonEmpty(value["id"]?.stringValue),
		      let projectCwd = nonEmpty(value["projectCwd"]?.stringValue)
		else { return nil }
		return RemoteSessionSummary(
			id: id,
			projectCwd: projectCwd,
			projectName: value["projectName"]?.stringValue ?? projectCwd,
			title: value["title"]?.stringValue ?? "",
			preview: value["preview"]?.stringValue,
			updatedAt: value["updatedAt"]?.numberValue ?? 0,
			status: readSessionStatus(value["status"]),
			live: value["live"]?.boolValue == true
		)
	}

	public static func readSessionSummaries(_ value: JSONValue?) -> [RemoteSessionSummary] {
		(value?["sessions"]?.arrayValue ?? []).compactMap(readSessionSummary)
	}

	public static func readQuestionRequest(_ value: JSONValue?) -> RemoteQuestionRequest? {
		guard let value, value.isObject, let requestId = nonEmpty(value["requestId"]?.stringValue),
		      let list = value["questions"]?.arrayValue
		else { return nil }
		let questions = list.filter(\.isObject).compactMap { item -> RemoteQuestionItem? in
			guard let question = nonEmpty(item["question"]?.stringValue) else { return nil }
			let options = (item["options"]?.arrayValue ?? []).filter(\.isObject).compactMap { option -> RemoteQuestionOption? in
				guard let label = nonEmpty(option["label"]?.stringValue) else { return nil }
				return RemoteQuestionOption(label: label, description: option["description"]?.stringValue ?? "")
			}
			return RemoteQuestionItem(
				question: question,
				header: item["header"]?.stringValue ?? "",
				options: options,
				multiSelect: item["multiSelect"]?.boolValue == true
			)
		}
		return RemoteQuestionRequest(requestId: requestId, questions: questions)
	}

	public static func readSessionState(_ value: JSONValue?) -> RemoteSessionState {
		guard let value, value.isObject else { return RemoteSessionState(status: .idle) }
		let error = value["error"].flatMap { error -> RemoteSessionError? in
			guard error.isObject else { return nil }
			return RemoteSessionError(
				code: error["code"]?.stringValue ?? "internal_error",
				message: error["message"]?.stringValue ?? ""
			)
		}
		return RemoteSessionState(
			status: readSessionStatus(value["status"]),
			detail: value["detail"]?.stringValue,
			model: value["model"]?.stringValue,
			contextPercent: value["contextPercent"]?.numberValue,
			error: error,
			pendingQuestion: readQuestionRequest(value["pendingQuestion"])
		)
	}

	public static func readToolEvent(_ value: JSONValue?) -> RemoteToolEvent? {
		guard let value, value.isObject,
		      let toolCallId = nonEmpty(value["toolCallId"]?.stringValue),
		      let toolName = nonEmpty(value["toolName"]?.stringValue),
		      let phase = value["phase"]?.stringValue.flatMap(RemoteToolPhase.init(rawValue:))
		else { return nil }
		return RemoteToolEvent(
			toolCallId: toolCallId,
			toolName: toolName,
			phase: phase,
			args: value["args"]?.stringValue,
			result: value["result"]?.stringValue,
			label: value["label"]?.stringValue,
			durationMs: value["durationMs"]?.numberValue
		)
	}

	public static func readMessageEvent(_ value: JSONValue?, now: () -> Double = WallClock.nowMs) -> RemoteMessageEvent? {
		guard let value, value.isObject else { return nil }
		switch value["kind"]?.stringValue {
		case "user":
			guard let text = value["text"]?.stringValue else { return nil }
			return .user(text: text, at: value["at"]?.numberValue ?? now())
		case "assistant_delta":
			return value["text"]?.stringValue.map(RemoteMessageEvent.assistantDelta)
		case "thinking_delta":
			return value["text"]?.stringValue.map(RemoteMessageEvent.thinkingDelta)
		case "turn_end":
			return .turnEnd(at: value["at"]?.numberValue ?? now())
		default:
			return nil
		}
	}

	public static func readTranscriptEntries(_ value: JSONValue?) -> [RemoteTranscriptEntry] {
		(value?["entries"]?.arrayValue ?? []).compactMap { entry -> RemoteTranscriptEntry? in
			guard entry.isObject, let id = nonEmpty(entry["id"]?.stringValue) else { return nil }
			switch entry["kind"]?.stringValue {
			case "user":
				return .user(id: id, text: entry["text"]?.stringValue ?? "", at: entry["at"]?.numberValue)
			case "assistant":
				let calls = (entry["toolCalls"]?.arrayValue ?? []).filter(\.isObject).compactMap { call -> RemoteToolCallSummary? in
					guard let toolCallId = nonEmpty(call["toolCallId"]?.stringValue),
					      let toolName = nonEmpty(call["toolName"]?.stringValue)
					else { return nil }
					return RemoteToolCallSummary(
						toolCallId: toolCallId,
						toolName: toolName,
						args: call["args"]?.stringValue,
						result: call["result"]?.stringValue,
						isError: call["isError"]?.boolValue == true,
						durationMs: call["durationMs"]?.numberValue
					)
				}
				return .assistant(
					id: id,
					text: entry["text"]?.stringValue ?? "",
					thinking: entry["thinking"]?.stringValue,
					toolCalls: calls,
					at: entry["at"]?.numberValue,
					error: entry["error"]?.stringValue
				)
			case "marker":
				return .marker(id: id, text: entry["text"]?.stringValue ?? "", at: entry["at"]?.numberValue)
			default:
				return nil
			}
		}
	}

	public static func readDeviceStatus(_ value: JSONValue?) -> RemoteDeviceStatus? {
		guard let value, value.isObject, let deviceName = nonEmpty(value["deviceName"]?.stringValue) else { return nil }
		return RemoteDeviceStatus(
			deviceName: deviceName,
			osLabel: value["osLabel"]?.stringValue,
			lanEndpoints: (value["lanEndpoints"]?.arrayValue ?? []).compactMap(\.stringValue),
			relayEnabled: value["relayEnabled"]?.boolValue == true,
			runningSessionCount: value["runningSessionCount"]?.numberValue ?? 0
		)
	}

	public static func readDevicePaired(_ value: JSONValue?) -> RemoteDevicePaired? {
		guard let value, value.isObject,
		      let pairingId = nonEmpty(value["pairingId"]?.stringValue),
		      let mobileSecret = nonEmpty(value["mobileSecret"]?.stringValue),
		      let desktopName = nonEmpty(value["desktopName"]?.stringValue)
		else { return nil }
		return RemoteDevicePaired(
			pairingId: pairingId,
			mobileSecret: mobileSecret,
			desktopName: desktopName,
			lanEndpoints: (value["lanEndpoints"]?.arrayValue ?? []).compactMap(\.stringValue),
			relayBaseUrl: value["relayBaseUrl"]?.stringValue
		)
	}

	public static func readProjectSummaries(_ value: JSONValue?) -> [RemoteProjectSummary] {
		(value?["projects"]?.arrayValue ?? []).compactMap { entry -> RemoteProjectSummary? in
			guard entry.isObject, let cwd = nonEmpty(entry["cwd"]?.stringValue) else { return nil }
			return RemoteProjectSummary(
				cwd: cwd,
				name: entry["name"]?.stringValue ?? cwd,
				kind: entry["kind"]?.stringValue == "conversation" ? "conversation" : "project",
				sessionCount: entry["sessionCount"]?.numberValue ?? 0
			)
		}
	}

	/// JS truthiness for strings: empty strings count as missing.
	private static func nonEmpty(_ text: String?) -> String? {
		guard let text, !text.isEmpty else { return nil }
		return text
	}
}

File: apps/mobile/client-android/shared/src/commonMain/kotlin/org/vetta/android/domain/remote/RemoteApi.kt
package org.vetta.android.domain.remote

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.put

/*
 * Payload contracts carried inside requests, responses and events (port of the
 * iOS `RemoteAPI.swift`, itself a port of `packages/remote-control/src/api.ts`).
 * Readers are tolerant: unknown fields are dropped and malformed entries skipped.
 */

@Serializable
enum class RemoteSessionStatus {
    @SerialName("idle") Idle,
    @SerialName("running") Running,
    @SerialName("thinking") Thinking,
    @SerialName("waiting_input") WaitingInput,
    @SerialName("completed") Completed,
    @SerialName("error") Error,
    @SerialName("aborted") Aborted,
    ;

    val isActive: Boolean
        get() = this == Running || this == Thinking || this == WaitingInput

    companion object {
        fun parse(value: String?): RemoteSessionStatus =
            when (value) {
                "running" -> Running
                "thinking" -> Thinking
                "waiting_input" -> WaitingInput
                "completed" -> Completed
                "error" -> Error
                "aborted" -> Aborted
                else -> Idle
            }
    }
}

@Serializable
data class RemoteProjectSummary(
    val cwd: String,
    val name: String,
    /** "conversation" for the desktop's project-less chats, otherwise "project". */
    val kind: String,
    val sessionCount: Int,
) {
    val isConversation: Boolean
        get() = kind == "conversation"
}

@Serializable
data class RemoteSessionSummary(
    val id: String,
    val projectCwd: String,
    val projectName: String,
    val title: String,
    val preview: String? = null,
    val updatedAt: Long,
    val status: RemoteSessionStatus,
    /** True when the desktop currently holds a live runtime instance for it. */
    val live: Boolean,
    /** Set while pinned, shared with the desktop sidebar; newer pins sort first. */
    val pinnedAt: Long? = null,
) {
    val pinned: Boolean
        get() = pinnedAt != null
}

@Serializable
data class RemoteQuestionOption(val label: String, val description: String)

@Serializable
data class RemoteQuestionItem(
    val question: String,
    val header: String,
    val options: List<RemoteQuestionOption>,
    val multiSelect: Boolean,
)

@Serializable
data class RemoteQuestionRequest(val requestId: String, val questions: List<RemoteQuestionItem>)

data class RemoteQuestionAnswer(val question: String, val answers: List<String>) {
    fun toJson(): JsonObject =
        buildJsonObject {
            put("question", question)
            put("answers", buildJsonArray { answers.forEach { add(JsonPrimitive(it)) } })
        }
}

/** What a `session.upload` carries: a picture the desktop can look at, or any other file. */
@Serializable
enum class AttachmentKind(val wire: String) {
    @SerialName("image") Image("image"),
    @SerialName("file") File("file"),
}

@Serializable
data class RemoteSessionError(val code: String, val message: String)

@Serializable
data class RemoteSessionState(
    val status: RemoteSessionStatus,
    val detail: String? = null,
    /** Display name of the session's model. */
    val model: String? = null,
    /** `provider/modelId`, as `session.configure` takes it. */
    val modelKey: String? = null,
    val thinkingLevel: String? = null,
    val contextPercent: Double? = null,
    val error: RemoteSessionError? = null,
    /** Present while the desktop waits for an answer that the phone may give. */
    val pendingQuestion: RemoteQuestionRequest? = null,
)

/** A model the session can switch to, with the thinking levels it accepts. */
@Serializable
data class RemoteModelOption(
    /** `provider/modelId`. */
    val key: String,
    val name: String,
    val provider: String,
    /** Empty when the model has no thinking control; otherwise starts with "off" or "none". */
    val thinkingLevels: List<String>,
    val defaultThinkingLevel: String? = null,
    val supportsImage: Boolean,
)

enum class RemoteToolPhase {
    Generating,
    Started,
    Updated,
    Phase,
    Completed,
    Failed,
    ;

    companion object {
        fun parse(value: String?): RemoteToolPhase? =
            when (value) {
                "generating" -> Generating
                "started" -> Started
                "updated" -> Updated
                "phase" -> Phase
                "completed" -> Completed
                "failed" -> Failed
                else -> null
            }
    }
}

data class RemoteToolEvent(
    val toolCallId: String,
    val toolName: String,
    val phase: RemoteToolPhase,
    val args: String? = null,
    val result: String? = null,
    val label: String? = null,
    val durationMs: Double? = null,
)

sealed interface RemoteMessageEvent {
    data class User(val text: String, val at: Long) : RemoteMessageEvent

    data class AssistantDelta(val text: String) : RemoteMessageEvent

    data class ThinkingDelta(val text: String) : RemoteMessageEvent

    data class TurnEnd(val at: Long) : RemoteMessageEvent
}

data class RemoteToolCallSummary(
    val toolCallId: String,
    val toolName: String,
    val args: String? = null,
    val result: String? = null,
    val isError: Boolean = false,
    val durationMs: Double? = null,
)

sealed interface RemoteTranscriptEntry {
    val id: String

    data class User(override val id: String, val text: String, val at: Long?) : RemoteTranscriptEntry

    data class Assistant(
        override val id: String,
        val text: String,
        val thinking: String?,
        val toolCalls: List<RemoteToolCallSummary>,
        val at: Long?,
        val error: String?,
    ) : RemoteTranscriptEntry

    data class Marker(override val id: String, val text: String, val at: Long?) : RemoteTranscriptEntry
}

data class RemoteDeviceStatus(
    val deviceName: String,
    val osLabel: String?,
    val lanEndpoints: List<String>,
    val relayEnabled: Boolean,
    val runningSessionCount: Int,
)

/** Sealed follow-up to a manual pairing approval; carries the long-lived credential. */
data class RemoteDevicePaired(
    val pairingId: String,
    val mobileSecret: String,
    val desktopName: String,
    val lanEndpoints: List<String>,
    val relayBaseUrl: String?,
)

object RemoteApi {
    /**
     * Largest attachment one `session.upload` may carry, before base64: one
     * sealed frame holds ~1 MB of JSON, so attachments travel one per request.
     */
    const val MAX_UPLOAD_BYTES = 700 * 1024

    fun readSessionSummary(value: JsonElement?): RemoteSessionSummary? {
        val obj = value as? JsonObject ?: return null
        val id = obj.text("id") ?: return null
        val projectCwd = obj.text("projectCwd") ?: return null
        return RemoteSessionSummary(
            id = id,
            projectCwd = projectCwd,
            projectName = obj.string("projectName") ?: projectCwd,
            title = obj.string("title").orEmpty(),
            preview = obj.string("preview"),
            updatedAt = obj.long("updatedAt") ?: 0,
            status = RemoteSessionStatus.parse(obj.string("status")),
            live = obj.bool("live") == true,
            pinnedAt = obj.long("pinnedAt"),
        )
    }

    fun readSessionSummaries(value: JsonElement?): List<RemoteSessionSummary> =
        (value as? JsonObject).array("sessions").mapNotNull(::readSessionSummary)

    fun readQuestionRequest(value: JsonElement?): RemoteQuestionRequest? {
        val obj = value as? JsonObject ?: return null
        val requestId = obj.text("requestId") ?: return null
        val list = obj["questions"] as? JsonArray ?: return null
        val questions =
            list.mapNotNull { item ->
                val question = item as? JsonObject ?: return@mapNotNull null
                val text = question.text("question") ?: return@mapNotNull null
                val options =
                    question.array("options").mapNotNull { option ->
                        val entry = option as? JsonObject ?: return@mapNotNull null
                        val label = entry.text("label") ?: return@mapNotNull null
                        RemoteQuestionOption(label, entry.string("description").orEmpty())
                    }
                RemoteQuestionItem(
                    question = text,
                    header = question.string("header").orEmpty(),
                    options = options,
                    multiSelect = question.bool("multiSelect") == true,
                )
            }
        return RemoteQuestionRequest(requestId, questions)
    }

    fun readSessionState(value: JsonElement?): RemoteSessionState {
        val obj = value as? JsonObject ?: return RemoteSessionState(RemoteSessionStatus.Idle)
        val error =
            (obj["error"] as? JsonObject)?.let {
                RemoteSessionError(code = it.string("code") ?: "internal_error", message = it.string("message").orEmpty())
            }
        return RemoteSessionState(
            status = RemoteSessionStatus.parse(obj.string("status")),
            detail = obj.string("detail"),
            model = obj.string("model"),
            modelKey = obj.string("modelKey"),
            thinkingLevel = obj.string("thinkingLevel"),
            contextPercent = obj.double("contextPercent"),
            error = error,
            pendingQuestion = readQuestionRequest(obj["pendingQuestion"]),
        )
    }

    fun readToolEvent(value: JsonElement?): RemoteToolEvent? {
        val obj = value as? JsonObject ?: return null
        return RemoteToolEvent(
            toolCallId = obj.text("toolCallId") ?: return null,
            toolName = obj.text("toolName") ?: return null,
            phase = RemoteToolPhase.parse(obj.string("phase")) ?: return null,
            args = obj.string("args"),
            result = obj.string("result"),
            label = obj.string("label"),
            durationMs = obj.double("durationMs"),
        )
    }

    fun readMessageEvent(value: JsonElement?, now: () -> Long): RemoteMessageEvent? {
        val obj = value as? JsonObject ?: return null
        return when (obj.string("kind")) {
            "user" -> RemoteMessageEvent.User(obj.string("text") ?: return null, obj.long("at") ?: now())
            "assistant_delta" -> obj.string("text")?.let(RemoteMessageEvent::AssistantDelta)
            "thinking_delta" -> obj.string("text")?.let(RemoteMessageEvent::ThinkingDelta)
            "turn_end" -> RemoteMessageEvent.TurnEnd(obj.long("at") ?: now())
            else -> null
        }
    }

    fun readTranscriptEntries(value: JsonElement?): List<RemoteTranscriptEntry> =
        (value as? JsonObject).array("entries").mapNotNull { element ->
            val entry = element as? JsonObject ?: return@mapNotNull null
            val id = entry.text("id") ?: return@mapNotNull null
            when (entry.string("kind")) {
                "user" -> RemoteTranscriptEntry.User(id, entry.string("text").orEmpty(), entry.long("at"))
                "assistant" ->
                    RemoteTranscriptEntry.Assistant(
                        id = id,
                        text = entry.string("text").orEmpty(),
                        thinking = entry.string("thinking"),
                        toolCalls =
                            entry.array("toolCalls").mapNotNull { element ->
                                val call = element as? JsonObject ?: return@mapNotNull null
                                RemoteToolCallSummary(
                                    toolCallId = call.text("toolCallId") ?: return@mapNotNull null,
                                    toolName = call.text("toolName") ?: return@mapNotNull null,
                                    args = call.string("args"),
                                    result = call.string("result"),
                                    isError = call.bool("isError") == true,
                                    durationMs = call.double("durationMs"),
                                )
                            },
                        at = entry.long("at"),
                        error = entry.string("error"),
                    )
                "marker" -> RemoteTranscriptEntry.Marker(id, entry.string("text").orEmpty(), entry.long("at"))
                else -> null
            }
        }

    fun readDeviceStatus(value: JsonElement?): RemoteDeviceStatus? {
        val obj = value as? JsonObject ?: return null
        return RemoteDeviceStatus(
            deviceName = obj.text("deviceName") ?: return null,
            osLabel = obj.string("osLabel"),
            lanEndpoints = obj.array("lanEndpoints").mapNotNull { (it as? JsonPrimitive)?.takeIf(JsonPrimitive::isString)?.content },
            relayEnabled = obj.bool("relayEnabled") == true,
            runningSessionCount = obj.double("runningSessionCount")?.toInt() ?: 0,
        )
    }

    fun readDevicePaired(value: JsonElement?): RemoteDevicePaired? {
        val obj = value as? JsonObject ?: return null
        return RemoteDevicePaired(
            pairingId = obj.text("pairingId") ?: return null,
            mobileSecret = obj.text("mobileSecret") ?: return null,
            desktopName = obj.text("desktopName") ?: return null,
            lanEndpoints = obj.array("lanEndpoints").mapNotNull { (it as? JsonPrimitive)?.takeIf(JsonPrimitive::isString)?.content },
            relayBaseUrl = obj.string("relayBaseUrl"),
        )
    }

    fun readProjectSummaries(value: JsonElement?): List<RemoteProjectSummary> =
        (value as? JsonObject).array("projects").mapNotNull { element ->
            val entry = element as? JsonObject ?: return@mapNotNull null
            val cwd = entry.text("cwd") ?: return@mapNotNull null
            RemoteProjectSummary(
                cwd = cwd,
                name = entry.string("name") ?: cwd,
                kind = if (entry.string("kind") == "conversation") "conversation" else "project",
                sessionCount = entry.double("sessionCount")?.toInt() ?: 0,
            )
        }

    fun readModelOptions(value: JsonElement?): List<RemoteModelOption> =
        (value as? JsonObject).array("models").mapNotNull { element ->
            val entry = element as? JsonObject ?: return@mapNotNull null
            val key = entry.text("key") ?: return@mapNotNull null
            RemoteModelOption(
                key = key,
                name = entry.text("name") ?: key,
                provider = entry.string("provider") ?: key.substringBefore('/'),
                thinkingLevels = entry.array("thinkingLevels").mapNotNull { (it as? JsonPrimitive)?.takeIf(JsonPrimitive::isString)?.content?.takeIf(String::isNotEmpty) },
                defaultThinkingLevel = entry.string("defaultThinkingLevel"),
                supportsImage = entry.bool("supportsImage") == true,
            )
        }
}

private fun JsonObject.string(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

/** JS truthiness for strings: empty strings count as missing. */
private fun JsonObject.text(key: String): String? = string(key)?.takeIf(String::isNotEmpty)

private fun JsonObject.double(key: String): Double? = (this[key] as? JsonPrimitive)?.takeUnless { it.isString }?.doubleOrNull

private fun JsonObject.long(key: String): Long? = double(key)?.toLong()

private fun JsonObject.bool(key: String): Boolean? = (this[key] as? JsonPrimitive)?.takeUnless { it.isString }?.booleanOrNull

private fun JsonObject?.array(key: String): JsonArray = this?.get(key) as? JsonArray ?: JsonArray(emptyList())
